### ISSUE TITLE
Galil: Be able to home on an inverted home switch

### ISSUE THREAD

**Issue by DominicOram** [ISISComputingGroup/IBEX#2962]
As an instrument scientist I would like to be able to home an axis to it's home datum regardless of the normally open/closed behaviour of the switch. Whilst this couldn't be done using the Galil HM command we should be able to write our own Galil code to do it. This may be a good opportunity to pair up with someone from motion control

**Comment by DominicOram** [comment]
FYI @jonelmer

**Comment by GDH-ISIS** [comment]
I think caution should be used here as the axis wiring would be inherently wrong - would it not be better to try to correct it?
Note : if a program was placed in the Galil or run from the computer, it might be victim of the polling rate. That said, it may be sufficient if the speeds are slow enough.

**Comment by DominicOram** [comment]
Yep, I'm aware its an axis wiring issue, stemming from a confusion over coordinates. In this case, I was told by Phil that the switch itself couldn't just be rewired,  it would require a small circuit to invert the signal or a complete replacement switch. Potentially that is the answer but we should discuss with motion control, which we could do whilst discussing #2963

**Comment by jonelmer** [comment]
FYI @IainJohnson

**Comment by IainJohnson** [comment]
Hello Dom,

There is a risk here of several issues becoming entangled, and IMHO some issues that are not issues. I'm more than hapy to talk these through with anyone, but I've had 3 people ask me about this in as many days, so I need to get these thoughts on paper just for now!;

Yes - I think writing bespoke code is probably the way to go.

**I've some observations:**

• The command for setting up home switch direction CN affects all 8 axes.
• The historic default of CN-1,-1 does not match the hardware layout on some recent designs.
*OR*
• The hardware layout on some recent designs does not match CN-1,-1.

:)

**Commands:**
The Galil has 3 inherant ways to home, and countless alternative possibilities.
FE for finding the edge of a home switch
FI for finding an Index Mark of an encoder
HM a simple routing for finding the home switch then the next index mark.

**Find Edge and Switch wiring:**
According to the Galil Manual, the FE command should work for any sense of switch. FE looks for *transition* of state. What's important is the switch position in the mechanical layout matches the *direction* criteria for CN-1,-1. I.e. the home switch needs to be EITHER Normally Closed AND positioned towards the Negative Limit, OR Normally Open AND positioned towards the positive limit.

NOTE Galil assumes after the FE move, you then go off and find an Index, or move to an offset position to finish. Galil does not assume you stay parked on the switch. (Imagine what happens if you do a FE search starting from being already parked on the switch?) The FE move on it's own is rubbery, finishes-up depending on speed and decelration.
Stage manufacturers such as Newport make stages that have a home switch with just ONE transition for the complete travel, and this is where the FE command is best suited. (Whereas our famous microswitch placed between limits has TWO transitions if you traverse across it.)

If we do our own thing, initiate a move in another way, such as PR JG etc, in the direction we want, then test the state of the home switch using MG_HMn, then the afore mentioned problems go away.

**My thoughts on wiring sense and failsafes:**
Home switches should not be used for hardware protection, i.e. not be used as limits. For zonal protection, a range switch should be used.
That said, then a home switch can be normally open or normally closed. My personal preference is normally open, this way if a wire falls off, the axis never homes, which gives you a clue something is wrong.

**Other Issues encountered recently:**
IDD & Computing & OEM direction XYZ co-ordinates do not match.
Limit Switch also triggers Home Switch - This is in the Galil Driver Layer and does not affect the low level behaviour. #2941

Cheers, Iain

**Comment by jonelmer** [comment]
Given the wild inconsistency and ambiguity inherent in the labelling of high/low limits on our motion racks, & drawings, and how unpredictably these are adhered to, I think it is pretty much a lucky dip as to which orientation the home flags are (as a vague observation, it's correct for about 50% of axes!). 

Having a further way to reverse the home behaviour would mean that the Galil system can cope with whatever hardware id provided, which I'm all for!

A further hardware parameter which I believe is worth investigating is reversing the motor wiring at the device, which *could potentially* allow the hardware to compensate for a wrongly oriented home flag and use the current homing routine.

**Comment by GDH-ISIS** [comment]
The homing routine for an inverted homing switch may look similar to the attached file. Appears to work.
[galil_Home_Inverted_Home.zip](https://github.com/ISISComputingGroup/IBEX/files/2097921/galil_Home_Inverted_Home.zip)
